sub-packages/bionemo-llm/src/bionemo/llm/train.py

```diff
@@ -0,0 +1,236 @@
+
+
+def nemo_logger_factory(experiment_config: ExperimentConfig, wandb_config: Optional[WandbConfig]) -> nl.NeMoLogger:
+    """Creates and returns a NeMoLogger instance configured based on the provided experiment and wandb configurations.
+
+    Args:
+        experiment_config (ExperimentConfig): Configuration object containing experiment settings such as
+            result directory, experiment name, checkpoint settings, and logger preferences.
+        wandb_config (Optional[WandbConfig]): Optional configuration object for Weights and Biases logging.
+
+    Returns:
+        nl.NeMoLogger: An instance of NeMoLogger configured with the specified settings.
+    """
+    checkpoint_callback = nl_callbacks.ModelCheckpoint(
+        save_last=experiment_config.save_last_checkpoint,
+        monitor=experiment_config.metric_to_monitor_for_checkpoints,
+        save_top_k=experiment_config.save_top_k,
+        every_n_train_steps=experiment_config.save_every_n_steps,
+        always_save_context=True,
+    )
+
+    nemo_logger = setup_nemo_lightning_logger(
+        root_dir=experiment_config.result_dir,
+        name=experiment_config.experiment_name,
+        initialize_tensorboard_logger=experiment_config.create_tensorboard_logger,
+        wandb_config=wandb_config,
+        ckpt_callback=checkpoint_callback,
+    )
+    return nemo_logger
+
+
+def setup_trainer(
+    parallel_config: ParallelConfig,
+    training_config: TrainingConfig,
+    callbacks=None,
+    nsys_config: NsysConfig | None = None,
+) -> nl.Trainer:
+    """Set up the trainer for model training using the specified parallel and training configurations.
+
+    Args:
+        parallel_config (ParallelConfig): Configuration for parallelism, including tensor and pipeline model parallel sizes,
+                                          number of devices, and number of nodes.
+        training_config (TrainingConfig): Configuration for training, including maximum steps, accelerator type,
+                                          validation batch limit, validation check interval, and precision.
+        callbacks (list, optional): List of callback functions to be used during training. Defaults to None,
+                                    in which case default callbacks (RichModelSummary and LearningRateMonitor) are used.
+        nsys_config (NsysConfig, optional): Configuration for nsys profiling. If None, is disabled.
+
+    Returns:
+        nl.Trainer: Configured trainer object ready for model training.
+    """
+    strategy = nl.MegatronStrategy(
+        tensor_model_parallel_size=parallel_config.tensor_model_parallel_size,
+        pipeline_model_parallel_size=parallel_config.pipeline_model_parallel_size,
+        ddp="megatron",
+        find_unused_parameters=True,
+        ckpt_include_optimizer=True,
+    )
+    if callbacks is None:
+        callbacks = [
+            RichModelSummary(max_depth=4),
+            LearningRateMonitor(),
+        ]
```

Comment:
Is there a way for users to add custom callback/loggers like perplexity logging?

Reply:
Yeah like ESM2 uses perplexity, geneformer does not. ESM2 uses a loss that passes enough info through to calculate perplexity at the cost of some forward pass overhead while geneformer uses a version with defaults that tell it to not compute that extra info.

Reply:
I added a hook for perplexity, but we will need to come up with something more stable. I think callbacks are somewhat unbounded similar to the scheduler, and we need a way to restrict this set for validation.

Reply:
ok done.